### PR TITLE
Added Normalization Guide.

### PR DESCRIPTION
**Pull request by thomelane**
Added python/guide/module/gluon/normalization.md and associated images.

### CODE REVIEW

**Comment by thomelane** [comment]
@Ishitori please review.

File: python/guide/module/gluon/normalization.md [hunk 1]
```diff
@@ -0,0 +1,412 @@
+# Normalization Blocks
+
+When training deep neural networks there are a number of techniques that are thought to be essential for model convergence. One important area is deciding how to initialize the parameters of the network. Using techniques such as Xavier initialization, we can can improve the gradient flow through the network at the start of training. Another important technique is normalization: i.e. scaling and shifting certain values towards a distribution with a mean of 0 (i.e. zero centered) and a standard distribution of 1 (i.e. unit variance). Which values you normalize depends on the exact method used as we'll see later on.
```
Comment (vishaalkapoor) [inline python/guide/module/gluon/normalization.md]: zero-centered

Comment (vishaalkapoor) [inline python/guide/module/gluon/normalization.md]: Suggestion: When training deep neural networks .... Model convergence significantly depends on how you initialize the parameters of the network. Using techniques such as ..... We can also use a technique called normalization: i.e.

File: python/guide/module/gluon/normalization.md [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+### Data Normalization
+
+One of the first applications of normalization is on the input data to the network. Although this whole section can be skipped in favor of an initial `BatchNorm` layer in your network, which will have a very similar effect (see next section).
+
+* **Step 1** is to calculate the mean and standard deviation of the entire training dataset. You'll usually want to do this for each channel separately. Sometimes you'll see normalization on images applied per pixel, but per channel is more common.
+* **Step 2** is to use these statistics to normalize each batch for training and for inference too.
+
+Warning: You should calculate the normalization means and standard deviations using the training dataset only. Any leakage of information from you testing dataset will effect the reliability of your testing metrics.
+
+When using pre-trained models from the Gluon Model Zoo you'll sometimes see the normalization statistics used for training. You'll want to use these statistics to normalize your own input data for fine-tuning or inference with these models. Using `transforms.Normalize` is one way of applying the normalization, and this should be used in the `Dataset`.
```
Comment (vishaalkapoor) [inline python/guide/module/gluon/normalization.md]: Is there a link to the Gluon Model Zoo?

File: python/guide/module/gluon/normalization.md [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+normalizer = Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
+image = normalizer(image_float)
+image
+```
+
+
+
+
+    
+    [[[[-1.31304061  1.17004883]
```
Comment (vishaalkapoor) [inline python/guide/module/gluon/normalization.md]: This will be generated and can be removed.

**Comment by vishaalkapoor** [comment]
LGTM modulo comments

File: python/guide/module/gluon/normalization.md [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+
+    beta: [ 0.  0.]
+    gamma: [ 1.  1.]
+
+
+##  Instance Normalization
+
+Another less common normalization technique is called `InstanceNorm`, which can be useful for certain tasks such as image stylization. Unlike `BatchNorm` which normalizes across all samples of a batch per channel, `InstanceNorm` normalizes **across all spatial dimensions per channel per sample** (i.e. each sample of a batch is normalized independently).
+
+Watch out: `InstanceNorm` is ideal is better suited to convolutional networks (CNNs) than recurrent networks (RNNs). We expect the input distribution to the recurrent cell to change over time, so normalization over time doesn't work well. LayerNorm is better suited for this case.
```
Comment (Ishitori) [inline python/guide/module/gluon/normalization.md]: Typo? _in_ ideal is?

File: python/guide/module/gluon/normalization.md (hunk 1 above)
Comment (thomelane) [reply python/guide/module/gluon/normalization.md]: updated.

File: python/guide/module/gluon/normalization.md (hunk 2 above)
Comment (thomelane) [reply python/guide/module/gluon/normalization.md]: updated.

File: python/guide/module/gluon/normalization.md (hunk 3 above)
Comment (thomelane) [reply python/guide/module/gluon/normalization.md]: updated.

File: python/guide/module/gluon/normalization.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/module/gluon/normalization.md]: updated.

File: python/guide/module/gluon/normalization.md [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+    <p align="center">Figure 1: Data Normalization
+        <a href="http://cs231n.github.io/neural-networks-2/">(Source)</a>
+    </p>
+</p>
+
+Why does this help? [Some research](https://papers.nips.cc/paper/7515-how-does-batch-normalization-help-optimization.pdf) has found that networks with normalization have a loss function that's easier to optimize using stochastic gradient descent. Other reasons are that it prevents saturation of activations and prevents certain features from dominating due to differences in scale.
+
+### Data Normalization
+
+One of the first applications of normalization is on the input data to the network. Although this whole section can be skipped in favor of an initial `BatchNorm` layer in your network, which will have a very similar effect (see next section).
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/module/gluon/normalization.md]: updated.

File: python/guide/module/gluon/normalization.md [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+When using pre-trained models from the Gluon Model Zoo you'll sometimes see the normalization statistics used for training. You'll want to use these statistics to normalize your own input data for fine-tuning or inference with these models. Using `transforms.Normalize` is one way of applying the normalization, and this should be used in the `Dataset`.
+
+
+```python
+import mxnet as mx
+from mxnet.gluon.data.vision.transforms import Normalize
+
+image_int = mx.nd.random.randint(low=0, high=256, shape=(1,3,2,2))
+image_float = image_int.astype('float32')/255
+normalizer = Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/module/gluon/normalization.md]: updated.

File: python/guide/module/gluon/normalization.md [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+## Batch Normalization
+
+Figure 1: `BatchNorm` on NCHW data | Figure 2: `BatchNorm` on NTC data 
+- | - 
+![alt](./imgs/NCHW_BN.png) | ![alt](./imgs/NTC_BN.png)
+(e.g. batch of images) using the default of `axis=1` | (e.g. batch of sequences) overriding the default with `axis=2` (or `axis=-1`)
+
+One of the most popular normalization techniques is batch normalization (often called BatchNorm). We normalize the activations **across all samples in a batch** for each of the channels independently. See Figure 1. We calculate two batch (or local) statistics for every channel to perform the normalization: the mean and variance of the activations in that channel for all samples in a batch. And we use these to shift and scale respectively.
+
+Tip: we can used this at the start of a network to perform data normalization, although this is not exactly equivalent to the example seen above.
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/module/gluon/normalization.md]: updated.

File: python/guide/module/gluon/normalization.md [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+    automatically calculated: [-1.32424402]
+
+
+As mentioned before, `BatchNorm` has a number of parameters that update throughout training. 2 of the parameters are not updated in the typical fashion (using gradients), but instead are updated deterministically using exponential smoothing. We need to keep track of the average mean and variance of batches during training, so that we can use these values for normalization during inference.
+
+Advanced: Often during inference, we have a batch size of 1 so batch variance would be impossible to calculate. And we might get a data distribution shift between training and inference data, which shouldn't just be normalized away.
+
+Warning: `use_global_stats=True` is a method of using the global running statistics during training, but this will prevent the global statistics from updating. Also, it has no effect on inference mode. Only use this for pre-trained models where the global statistics have been calculated already and you want to use them during training. One example would be using a pre-trained ResNet as a image feature extractor in an Instance Segmentation model that can only be trained with small batches.
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/module/gluon/normalization.md]: updated.

File: python/guide/module/gluon/normalization.md [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+print("automatically calculated:", output[0][0][0][0].asnumpy())
+```
+
+    manually calculated: [-1.3242445]
+    automatically calculated: [-1.32424402]
+
+
+As mentioned before, `BatchNorm` has a number of parameters that update throughout training. 2 of the parameters are not updated in the typical fashion (using gradients), but instead are updated deterministically using exponential smoothing. We need to keep track of the average mean and variance of batches during training, so that we can use these values for normalization during inference.
+
+Advanced: Often during inference, we have a batch size of 1 so batch variance would be impossible to calculate. And we might get a data distribution shift between training and inference data, which shouldn't just be normalized away.
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/module/gluon/normalization.md]: updated.

File: python/guide/module/gluon/normalization.md [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+    running_vars: [ 17.24960899  17.24960899]
+
+
+#### Beta and Gamma
+
+As mentioned previously, there are two additional parameters in `BatchNorm` which are trainable in the typical fashion (with gradients). `beta` is used to shift and `gamma` is used to scale the normalized distribution, which allows the network to 'undo' the effects of normalization if required. You can prevent `beta` shifting by setting `center=False`, and prevent `gamma` scaling by setting `scale=False` when initializing the `BatchNorm` block.
+
+Watch out: zero centering and scaling to unit variance will still occur when setting `center=False` and `scale=False`. Only post normalization shifting and scaling will prevented.
+
+Advanced: although the `beta` and `gamma` parameters will still exist, they won't update (because their `grad_req=null`).
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/module/gluon/normalization.md]: updated.

File: python/guide/module/gluon/normalization.md [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+
+Some of the disadvantages of `BatchNorm` no longer apply. Small batch sizes are no longer an issue, since normalization statistics are calculated on single samples. And confusion around training and inference modes disappears because `LayerNorm` is the same for both modes.
+
+Warning: similar to having a small batch sizes in `BatchNorm`, you may have issues with `LayerNorm` if the input channel size is small. Using embeddings with a large enough dimension size avoids this (approx >20).
+
+Warning: currently MXNet Gluon's implementation of `LayerNorm` is applied along a single axis (which should be the channel axis). Other frameworks have the option to apply normalization across multiple axes, which leads to differences in `LayerNorm` on NCHW input by default. See Figure 3. Other frameworks can normalize over C, H and W, not just C as with MXNet Gluon.
+
+Remember: `LayerNorm` is intended to be used with data in NTC format so the default normalization axis is set to -1 (corresponding to C for channel). Change this to `axis=1` if you need to apply `LayerNorm` to data in NHWC format.
```
[the remark this answers is not in the document]
Comment (thomelane) [reply python/guide/module/gluon/normalization.md]: updated.

File: python/guide/module/gluon/normalization.md (hunk 4 above)
Comment (thomelane) [reply python/guide/module/gluon/normalization.md]: updated.

**Comment by thomelane** [comment]
Made adjustments based on Sergey's comments.
